### PR TITLE
Prometheus metrics

### PR DESCRIPTION
**Pull request by huntergregory**
This PR adds base Prometheus functionality for measuring and emitting custom metrics (time series) in NPM. There are 6 metrics included currently, and others can be added easily. The current metrics are: 
- number of network policies on the host
- execution time for adding a network policy
- number of IP Table rules on the host
- execution time for adding an IP Table rule to a chain on the host
- number of IP Sets on the host
- execution time for adding an IP Set to the host

Metrics are useful for troubleshooting network issues/configuration and are important to monitor (e.g. number of policies should remain low enough).

```
NPM Prometheus Metrics
```

### CODE REVIEW

**Comment by csfmomo** [comment]
Is there anyway to fix the following unit tests?

image

File: npm/metrics/http.go [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+	} else {
+		http.ListenAndServe(httpPort, nil)
+	}
+}
+
+// getHandler returns the HTTP handler for the metrics endpoint
+func getHandler() http.Handler {
+	if handler == nil {
+		handler = promhttp.Handler()
+		// 	handler = promhttp.HandlerFor(networkingRegistry, promhttp.HandlerOpts{}) // promhttp.Handler()
```
Comment (csfmomo) [inline npm/metrics/http.go:41]: Do we need this commented out code?

File: npm/metrics/http.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var started = false
+
+const httpPort = ":8000"
+
+var handler http.Handler
```
Comment (csfmomo) [inline npm/metrics/http.go:19]: Let's move it to line 15th, put all var together.

File: npm/metrics/http.go (hunk 1 above)
Comment (huntergregory) [reply npm/metrics/http.go:43]: just an alternative option, but not necessary. I'll delete it

**Comment by huntergregory** [comment]
> Is there anyway to fix the following unit tests?
> 
> image

Is there any reason why this wouldn't work in the pipeline environment: 
`http.Get("http://localhost" + metrics.HTTPPort + metrics.MetricsPath)`

**Comment by csfmomo** [comment]
Had an offline discussion with huntergregory for the following 4 topics. Topic 4 will be addressed in the following PR.
1. How to differentiate differentiate customer's data generated in Prometheus.
2. Who will have the access to Prometheus data.
3. How to isolate Prometheus data to limited audience.
4. Move test only code to testing directory or util directory. Separate testing code from prod code.

File: npm/metrics/test-util.go [hunk 3]
```diff
@@ -0,0 +1,22 @@
+package metrics
```
Comment (neaggarwMS) [inline npm/metrics/test-util.go:1]: Can you rename this file as util_test.go (and change format for all test files as <name>_test.go)?

File: npm/metrics/promutil/util.go [hunk 4]
```diff
@@ -0,0 +1,57 @@
+package promutil
```
Comment (neaggarwMS) [inline npm/metrics/promutil/util.go:1]: Can we move it out of promutil folder? or move its test file (test-util.go) in this folder. Usually test files remain next to the main files.

File: npm/metrics/testing/test-main.go [hunk 5]
```diff
@@ -0,0 +1,60 @@
+package main
```
Comment (neaggarwMS) [inline npm/metrics/testing/test-main.go:1]: Can you move test-main.go (rename it to main_test.go) out of testing folder.

File: npm/metrics/prometheus-metrics.go [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+func init() {
+	for metric := range allMetricNames {
+		prometheus.DefaultRegisterer.MustRegister(metric)
+	}
+}
+
+// Observe records a value in the given summary
+func Observe(summary prometheus.Summary, value float64) {
+	summary.Observe(value)
+	// if changed to a vector, use summary.WithLabelValues(hostName).Observe(value)
```
Comment (neaggarwMS) [inline npm/metrics/prometheus-metrics.go:67]: Is this comment relevant? if no, can we remove it?

File: npm/metrics/http.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+// Set asGoRoutine to true if you want to be able to effectively run other code after calling this.
+// The function will pause for delayAmountAfterStart seconds after starting the HTTP server for the first time.
+func StartHTTP(asGoRoutine bool, delayAmountAfterStart int) {
+	if started {
+		return
+	}
+	started = true
+
+	http.Handle(MetricsPath, getHandler())
+	if asGoRoutine {
```
Comment (neaggarwMS) [inline npm/metrics/http.go:31]: Shouldnt this be the only option?

File: npm/metrics/http.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+	//MetricsPath is the path for the Prometheus metrics endpoint (includes preceding slash)
+	MetricsPath = "/metrics"
+)
+
+var started = false
+var handler http.Handler
+
+// StartHTTP starts a HTTP server with endpoint on port 8000. Metrics are exposed on the endpoint /metrics.
+// Set asGoRoutine to true if you want to be able to effectively run other code after calling this.
+// The function will pause for delayAmountAfterStart seconds after starting the HTTP server for the first time.
```
Comment (neaggarwMS) [inline npm/metrics/http.go:23]: Can you explain why?

File: npm/plugin/main.go [hunk 9]
```diff
@@ -67,5 +68,7 @@ func main() {
 		panic(err.Error)
 	}
 
+	metrics.StartHTTP(false, 0)
```
Comment (neaggarwMS) [inline npm/plugin/main.go:71]: Shouldnt this be the firt call (with a go routine) before we start processing the rules already?

File: npm/nwpolicy.go [hunk 10]
```diff
@@ -146,6 +151,7 @@ func (npMgr *NetworkPolicyManager) DeleteNetworkPolicy(npObj *networkingv1.Netwo
 		err   error
 		ns    *namespace
 		allNs = npMgr.nsMap[util.KubeAllNamespacesFlag]
+		// timer = metrics.StartNewTimer()
```
Comment (neaggarwMS) [inline npm/nwpolicy.go:154]: Why is this deleted?

File: npm/iptm/iptm.go [hunk 11]
```diff
@@ -298,6 +299,8 @@ func (iptMgr *IptablesManager) DeleteChain(chain string) error {
 
 // Add adds a rule in iptables.
 func (iptMgr *IptablesManager) Add(entry *IptEntry) error {
+	timer := metrics.StartNewTimer()
```
Comment (neaggarwMS) [inline npm/iptm/iptm.go:302]: Can you also run your change against the load test. Please sync with Shufang for how to run the scale tests?

File: npm/metrics/promutil/util.go (hunk 4 above)
Comment (huntergregory) [reply npm/metrics/promutil/util.go:1]: test-util.go has utilities for testing. I'll move them into this file because this file is also utilities for testing

File: npm/plugin/main.go (hunk 9 above)
Comment (huntergregory) [reply npm/plugin/main.go:71]: The metrics are recorded in the same way whether or not they're available on an HTTP server. Having it last only means scrapers can't see the metrics until NPM is finished booting up.

File: npm/nwpolicy.go (hunk 10 above)
Comment (huntergregory) [reply npm/nwpolicy.go:154]: I was thinking of adding a metric for execution time for deleting network policies, ipsets, etc. later. It's super easy to add, so I'll just delete the comments for now.

File: npm/metrics/testing/test-main.go (hunk 5 above)
Comment (huntergregory) [reply npm/metrics/testing/test-main.go:1]: @neaggarwMS a main package has to be in a unique folder, and this file doesn't really test anything. It's more for displaying the metrics you've created on a localhost (outside the realm of kubernetes). I'll give it a more descriptive name

**Comment by huntergregory** [comment]
Noting here that the PR passed the local environment scale test (TH test). No errors and approximately the same memory usage in azure-npm pod (14 MiB without Prometheus and 18 MiB with Prometheus).

**Comment by huntergregory** [comment]
Letting @vishiy view the PR

File: npm/iptm/iptm.go (hunk 11 above)
Comment (neaggarwMS) [inline npm/iptm/iptm.go:302]: Did you run it with the load test?

File: npm/metrics/promutil/util.go (hunk 4 above)
Comment (neaggarwMS) [reply npm/metrics/promutil/util.go:1]: If this is only for testing, then can you rename it test-util.go. does it need to be inside promutil?
